Sources/SwiftBSON/BSONDocumentIterator.swift

```diff
@@ -95,126 +96,197 @@ public class BSONDocumentIterator: IteratorProtocol {
+            return self.moveReaderIndexSafely(forwardBy: 4)
 
         case .string, .code, .symbol:
             guard let strLength = buffer.readInteger(endianness: .little, as: Int32.self) else {
-                throw BSONError.InternalError(message: "Failed to read encoded string length")
+                return false
             }
-            self.buffer.moveReaderIndex(forwardBy: Int(strLength))
+            return self.moveReaderIndexSafely(forwardBy: Int(strLength))
 
         case .regex:
-            _ = try self.buffer.readCString()
-            _ = try self.buffer.readCString()
+            do {
+                _ = try self.buffer.readCString()
+                _ = try self.buffer.readCString()
+            } catch {
+                return false
+            }
 
         case .binary:
             guard let dataLength = buffer.readInteger(endianness: .little, as: Int32.self) else {
-                throw BSONError.InternalError(message: "Failed to read encoded binary data length")
+                return false
             }
-            self.buffer.moveReaderIndex(forwardBy: Int(dataLength) + 1) // +1 for the binary subtype.
+            return self.moveReaderIndexSafely(forwardBy: Int(dataLength) + 1) // +1 for the binary subtype.
 
         case .document, .array, .codeWithScope:
             guard let embeddedDocLength = buffer.readInteger(endianness: .little, as: Int32.self) else {
-                throw BSONError.InternalError(message: "Failed to read encoded document length")
+                return false
             }
             // -4 because the encoded length includes the bytes necessary to store the length itself.
-            self.buffer.moveReaderIndex(forwardBy: Int(embeddedDocLength) - 4)
+            return self.moveReaderIndexSafely(forwardBy: Int(embeddedDocLength) - 4)
 
         case .dbPointer:
             // initial string
             guard let strLength = buffer.readInteger(endianness: .little, as: Int32.self) else {
-                throw BSONError.InternalError(message: "Failed to read encoded string length")
+                return false
             }
-            self.buffer.moveReaderIndex(forwardBy: Int(strLength))
-            // 12 bytes of data
-            self.buffer.moveReaderIndex(forwardBy: 12)
+            return self.moveReaderIndexSafely(forwardBy: Int(strLength) + 12)
 
         case .decimal128:
-            self.buffer.moveReaderIndex(forwardBy: 16)
+            return self.moveReaderIndexSafely(forwardBy: 16)
         }
+
+        return true
     }
 
     /// Finds the key in the underlying buffer, and returns the [startIndex, endIndex) containing the corresponding
     /// element.
-    internal static func findByteRange(for searchKey: String, in document: BSONDocument) throws -> Range<Int>? {
+    /// Returns nil if invalid BSON is encountered.
+    internal static func findByteRange(for searchKey: String, in document: BSONDocument) -> Range<Int>? {
```

Comment:
Since this wasn't required by the benchmarks, I didn't originally update this to use the fast / unsafe `findValue`. It seems like a natural fit so I've updated it now.